crates/wasmparser/src/validator/core/canonical.rs

```diff
@@ -162,6 +170,111 @@ pub(crate) trait InternRecGroup {
         Ok(())
     }
 
+    fn check_descriptors(
+        &mut self,
+        rec_group: RecGroupId,
+        id: CoreTypeId,
+        types: &TypeList,
+        offset: usize,
+    ) -> Result<()> {
+        let ty = &types[id].composite_type;
+        if ty.descriptor_idx.is_some() || ty.describes_idx.is_some() {
+            if !self.features().custom_descriptors() {
+                return Err(BinaryReaderError::new(
+                    "custom descriptors proposal must be enabled to use descriptor and describes",
+                    offset,
+                ));
+            }
+            match &ty.inner {
+                CompositeInnerType::Struct(_) => (),
+                _ => {
+                    return Err(BinaryReaderError::new(
+                        if ty.descriptor_idx.is_some() {
+                            "descriptor clause on non-struct type"
+                        } else {
+                            "describes clause on non-struct type"
+                        },
+                        offset,
+                    ));
+                }
+            }
+        }
+
+        let map_cannonical = |idx: PackedIndex| -> Result<CoreTypeId> {
+            self.at_packed_index(types, rec_group, idx, offset)
+        };
+
+        let descriptor_idx = if let Some(i) = ty.descriptor_idx {
+            Some(map_cannonical(i)?)
+        } else {
+            None
+        };
+        let describes_idx = if let Some(i) = ty.describes_idx {
+            Some(map_cannonical(i)?)
+        } else {
+            None
+        };
+
+        if let Some(supertype_index) = types[id].supertype_idx {
+            debug_assert!(supertype_index.is_canonical());
+            let sup_id = map_cannonical(supertype_index)?;
+            if descriptor_idx.is_some() {
```

Comment:
Where possible I'd recommend changing this for example to `if let Some(descriptor_idx) = descriptor_idx` since that avoids `descriptor_idx.unwrap()` below within this scope. That's not possible everywhere below without Rust's relatively new stabilized feature of "let chains" but this clause for example should be easy to change